Adding new Object Display topic (#9726)
MystenLabs/sui: doc/src/build/sui-object-display.md

---
title: Sui Object Display Standard
---

> The Sui Object Display standard is a template engine that allows for off-chain management of display configurations for a type. With it, you can substitute data for an object into a template string. The standard doesn’t limit the fields you can set. You can use the  `{property}` syntax to access all object properties, and then insert them as a part of the template string.

Reviewer: Its should be a reverse - on-chain management for off-chain representation (display).

> You can use the `sui::display` module to define the display properties for a `Publisher` object you own. For more information about `Publisher` objects, see [Publisher](https://examples.sui.io/basics/publisher.html) topic in *Sui Move by Example*.

Reviewer: The opposite - `Publisher` object is required to set `sui::display` for a type.

In Sui Move, `Display<T>` represents an object that specifies a set of named templates for the type `T`. For example, for a type `0x2::capy::Capy` the display syntax is:  `Display<0x2::capy::Capy>`. 

> Sui Full nodes process all objects of the type `T` by matching the `Display` definition, and return the processed result when you query an object.

Reviewer: When query an object with `{ showDisplay: true }` setting in the query.

        );

        // Commit first version of `Display` to apply changes.
        display::update_version(&mut display);

        transfer(publisher, sender(ctx));
        transfer(display, sender(ctx));
    }

    /// Anyone can mint their `Hero`!
    public fun mint(name: String, img_url: String, ctx: &mut TxContext): Hero {
        let id = object::new(ctx);
        Hero { id, name, img_url }
    }
}
```

## Work with Object Display

The `display::new<T>` call creates a `Display`, either in a custom function or module initializer, or as part of a programmable transaction.
The following code sample demonstrates how to create a `Display`:

```rust
module sui::display {
    /// Get a new Display object for the `T`.
    /// Publisher must be the publisher of the T, `from_package`
    /// check is performed.
    public fun new<T>(pub: &Publisher): Display<T> { /* ... */ }
}
```

After you create the `Display`, you can modify it. The following code sample demonstrates how to modify a `Display`:

```rust
module sui::display {
    /// Sets multiple fields at once
    public fun add_multiple(
        self: &mut Display,
        keys: vector<String>,

>         values: vector<String

Reviewer: Something that we got wrong in the examples and needs fixing there as well.
```suggestion
        values: vector<String>
```

    ) { /* ... */ }

    /// Edit a single field
    public fun edit(self: &mut Display, key: String, value: String) { /* ... */ }

    /// Remove a key from Display
    public fun remove(self: &mut Display, key: String ) { /* ... */ }
}
```

Next, the `update_version` call applies the changes and sets the `Display` for the `T` by emitting an event. Full nodes receive the event and use the data in the event to retrieve a template for the type.

The following code sample demonstrates how to use the `update_version` call:

```rust
module sui::display {
    /// Update the version of Display and emit an event
    public fun update_version(self: &mut Display) { /* ... */ }
}
```

## Sui utility objects

> In Sui, utility objects enable authorization for Capabilities. Almost all modules have features that can be accessed only with the required Capability. Generic modules allow one Capability per Application, such as a marketplace. Some Capabilities mark ownership of a shared object on-chain, or access the shared data from another account.

Author: updated, ty, I didn't change the content from source doc

With capabilities, it is important to provide a meaningful description of objects to facilitate user interface implementation. This helps avoid accidentally transferring the wrong object when objects are similar. It also provides a user-friendly description of items that users see.

The following example demonstrates how to create a capy capability:

```rust
module capy::utility {
   /// A capability which grants Capy Manager permission to add
   /// new genes and manage the Capy Market
   struct CapyManagerCap has key, store {

> id: UID }

Reviewer: ```suggestion
        id: UID
    }
```

}
```

## Typical objects with data duplication 

A common case with in-game items is a large number of similar objects grouped by some criteria. It is important to optimize their size and the cost to mint and update them. Typically, a game uses a single source image or URL per group or item criteria. Storing the source image inside of every object is not optimal. 

> In some cases, in-game items are minted by users when a game allows them or when they purchase an in-game item. To enable this, some IPFS/Arweave metadata must be created and stored in advance. This requires additional logic that is usually not related to the in-game properties of the item.

Reviewer 2: ```suggestion
In some cases, users mint in-game items when a game allows them or when they purchase an in-game item. To enable this, some IPFS/Arweave metadata must be created and stored in advance. This requires additional logic that is usually not related to the in-game properties of the item.
```
Passive voice but not sure who is doing the creating and storing of metadata.

The following example demonstrates how to create a Capy:

```rust
module capy::capy_items {

>    /// A wearable Capy Item. For some items there can be an

Reviewer 2: ```suggestion
   /// A wearable Capy item. For some items there can be an
```

>    /// unlimited supply. And items with the same name are identical

Reviewer 2: ```suggestion
   /// unlimited supply. And items with the same name are identical.
```

   struct CapyItem has key, store {

> id: UID,

Reviewer: ```suggestion
        id: UID,
```

       name: String
   }
}
```

## Unique objects with dynamic representation

Sui Capys use dynamic image generation. When a Capy is born, its attributes determine the Capy’s appearance, such as color or pattern. When a user puts an item on a Capy, the Capy’s appearance changes. When users put multiple items on a Capy, there’s a chance of a bonus for a combination of items.

To implement this, the Capys game API service refreshes the image in response to a user-initiated change. The URL for a Capy is a template with the `capy.id`. But storing the full URL - as well as other fields in the Capy object due to their diverse population - also leads to users paying for excess storage and increased gas fees. 

> The following example demonstrates to implement dynamic image generation:

Reviewer 2: ```suggestion
The following example demonstrates how to implement dynamic image generation:
```

```rust
module capy::capy {
   /// A Capy - very diverse object with different combination
   /// of genes. Created dynamically + for images a dynamic SVG
   /// generation is used.
   struct Capy has key, store {

> id: UID,

Reviewer: ```suggestion
        id: UID,
```
